### PR TITLE
Add RegExp.prototype[@@split]

### PR DESCRIPTION
**Pull request by goyakin**

### CODE REVIEW

**Comment by goyakin** [comment]
@abchatra & @akroshg, could you take a look?

File: lib/Runtime/Library/JavascriptRegularExpression.cpp [hunk 1]
```diff
@@ -765,11 +817,46 @@ namespace Js
...
+        // if the algorithm is observable. If it is, we go through the new ES6 algorithm, but otherwise, we
+        // run the faster ES5 version.
+        //
+        // According to the spec, we're supposed to process "limit" after we use some of the RegExp properties.
+        // However, there doesn't seem to be any reason why "limit" processing can't be pulled above the rest
+        // in the spec. Therefore, we should see if such a spec update is OK. If not, this would have to be
+        // moved to its correct place in the code.
+        uint32 limit = (args.Info.Count < 3 || JavascriptOperators::IsUndefinedObject(args[2], scriptContext))
+            ? UINT_MAX
+            : JavascriptConversion::ToUInt32(args[2], scriptContext);
```
Comment (goyakin) [inline lib/Runtime/Library/JavascriptRegularExpression.cpp]: @bterlson This is what I talked to you about regarding where the `limit` argument is processed.

**Comment by goyakin** [comment]
Ping...

File: lib/Runtime/Base/ScriptContext.cpp [hunk 2]
```diff
@@ -850,17 +850,12 @@ namespace Js
         this->weakReferenceDictionaryList.Prepend(this->GeneralAllocator(), weakReferenceDictionary);
     }
 
-    RecyclableObject *ScriptContext::GetMissingPropertyResult(Js::RecyclableObject *instance, Js::PropertyId id)
+    RecyclableObject *ScriptContext::GetMissingPropertyResult()
```
Comment (akroshg) [inline lib/Runtime/Base/ScriptContext.cpp]: This is good.
While you are here - could you check and remove unwanted params for GetNullPropertyResult and GetNullItemResult below?

File: lib/Runtime/Library/JavascriptLibrary.cpp [hunk 3]
```diff
@@ -3890,30 +3893,26 @@ namespace Js
 
         if (scriptConfig->IsES6RegExPrototypePropertiesEnabled())
         {
-            RuntimeFunction* globalGetter = library->CreateGetterFunction(PropertyIds::global, &JavascriptRegExp::EntryInfo::GetterGlobal);
-            library->AddAccessorsToLibraryObject(regexPrototype, PropertyIds::global, globalGetter, nullptr);
-            library->regexGlobalGetterFunction = globalGetter;
-
+            library->regexGlobalGetterFunction =
```
Comment (akroshg) [inline lib/Runtime/Library/JavascriptLibrary.cpp]: Current function InitializeRegexPrototype should have passed right initSlotCapacity (typeHandler->Convert(, , 8) )?

File: lib/Runtime/Library/RegexHelper.cpp [hunk 4]
```diff
@@ -1213,7 +1212,156 @@ namespace Js
     }
 
     // String.prototype.split (ES5 15.5.4.14)
-    Var RegexHelper::RegexSplitImpl(ScriptContext* scriptContext, JavascriptRegExp* regularExpression, JavascriptString* input, CharCount limit, bool noResult, void *const stackAllocationPointer)
+    Var RegexHelper::RegexSplitImpl(ScriptContext* scriptContext, RecyclableObject* thisObj, JavascriptString* input, CharCount limit, bool noResult, void *const stackAllocationPointer)
```
Comment (akroshg) [inline lib/Runtime/Library/RegexHelper.cpp]: is above spec reference valid?

File: test/es6/regex-symbols.js [hunk 5]
```diff
@@ -600,9 +663,184 @@ var tests = [
...
+    },
+    {
+        name: "RegExp.prototype[@@split] should return an empty Array when the input size is 0 and the RegExp matches the empty string",
+        body: function () {
+            function assertResult(result) {
+                assert.areEqual(0, result.length, "result.length");
+            }
+            var re = /(?:)/;
+            var input = "";
+            verifySymbolSplitResult(assertResult, re, input);
```
Comment (akroshg) [inline test/es6/regex-symbols.js]: Did you have case where input and regex both are empty string ?

File: lib/Runtime/Library/RegexHelper.cpp [hunk 6]
```diff
@@ -1813,17 +1960,24 @@ namespace Js
...
+    CharCount RegexHelper::AdvanceStringIndex(JavascriptString* string, CharCount index, bool isUnicode)
     {
         // TODO: Change the increment to 2 depending on the "unicode" flag and
         // the code point at "index". The increment is currently constant at 1
         // in order to be compatible with the rest of the RegExp code.
-        return index + 1;
+
+        if (index + 1 < index) // Overflow
+        {
+            return MaxCharCount;
```
Comment (akroshg) [inline lib/Runtime/Library/RegexHelper.cpp]: we should throw here - Js::Throw::OutOfMemory();

File: test/es6/regex-symbols.js (hunk 5 above)
Comment (akroshg) [inline test/es6/regex-symbols.js]: Ok I meant new RegExp("")[Symbol.split](""), which is the same above case.

---
In reply to: [54801523](https://github.com/Microsoft/ChakraCore/pull/416#discussion_r54801523) [](ancestors = 54801523)

**Comment by akroshg** [comment]
Looks good to me.

File: lib/Runtime/Base/ScriptContext.cpp (hunk 2 above)
Comment (goyakin) [inline lib/Runtime/Base/ScriptContext.cpp]: It turns out we don't use these methods at all, so I removed them completely.

File: lib/Runtime/Library/JavascriptLibrary.cpp (hunk 3 above)
Comment (goyakin) [inline lib/Runtime/Library/JavascriptLibrary.cpp]: You're right, the initial slot capacity is wrong here. I changed it to `16` though, including the properties behind the `ES6RegExPrototypeProperties` flag.

File: lib/Runtime/Library/RegexHelper.cpp (hunk 4 above)
Comment (goyakin) [inline lib/Runtime/Library/RegexHelper.cpp]: I created a new function called `RegexEs5SplitImpl` for the ES5 version, but looks like I forgot to move the spec reference :smile:

File: lib/Runtime/Library/RegexHelper.cpp (hunk 6 above)
Comment (goyakin) [inline lib/Runtime/Library/RegexHelper.cpp]: I think it's better to cap the index at `MaxCharCount` instead. The new value is used for `lastIndex` and the next call to `exec` would result in **no** match, which would be the same as if we don't cap it.

**Comment by goyakin** [comment]
@dotnet-bot test Copyright Check please

**Comment by goyakin** [comment]
Thanks @akroshg for the review.

**Pull request merged**